lib/verdict/storage/redis_storage.rb

```diff
@@ -9,26 +9,30 @@ def initialize(redis = nil, options = {})
       end
 
       def get(scope, key)
-        redis.hget("#{@key_prefix}#{scope}", key)
+        redis.hget(scope_key(scope), key)
       rescue ::Redis::BaseError => e
         raise Verdict::StorageError, "Redis error: #{e.message}"
       end
 
       def set(scope, key, value)
-        redis.hset("#{@key_prefix}#{scope}", key, value)
+        redis.hset(scope_key(scope), key, value)
       rescue ::Redis::BaseError => e
         raise Verdict::StorageError, "Redis error: #{e.message}"
       end
 
       def remove(scope, key)
-        redis.hdel("#{@key_prefix}#{scope}", key)
+        redis.hdel(scope_key(scope), key)
       rescue ::Redis::BaseError => e
         raise Verdict::StorageError, "Redis error: #{e.message}"
       end
 
+      def cleanup(scope)
+        redis.del(scope_key(scope))
```

Comment:
The reason I never implemented this is that DEL on a hash has O(n) complexity where n is the number of hash entries (see https://redis.io/commands/del), which can be a really painful operation for large experiments, and has caused downtime in the past.

Reply:
You would need to use `HSCAN` (https://redis.io/commands/scan), and delete individual entries with HDEL, only to finally remove the hash with DEL once it is empty. During this process, you have to control the number of redis commands to not overload it.

Talk to the Redis/Memcache team for more info.

Reply:
I'm surprised we would ever be worried about calling. Docs say it is `O(n)`, where `n` is the number of keys being deleted. How many keys would we be deleting at once?

@shivnagarajan do you recall any issues about deleting keys in redis?

Reply:
> How many keys would we be deleting at once?

Some experiments may have assignments numbering in the hundreds of thousands.

Reply:
From https://redis.io/commands/del:
> When a key to remove holds a value other than a string, the individual complexity for this key is O(M) where M is the number of elements in the list, set, sorted set or hash.

Given that these hashes can contain hundreds of thousands of not millions of entries, this can make the redis instance unavailable for a long time. We had outages around Redis-maggedon due to this, and IIRC we moved the storage backend of experiments to MySQL because of this issue.

Reply:
>...hashes can contain hundreds of thousands if not millions of entries

Yeah, that won't work as a single operation.

Reply:
Yeah, I think the `HSCAN`, followed by `HDEL` seems like the best way forward here.